### DISCUSSION TITLE
Help Getting Started Programming Mastodon Profile Page

### DISCUSSION THREAD

**Discussion by jtayler** [mastodon/mastodon#28058] [Q&A]
First time programming Mastodon

I need to insert some interface based on data called from an API.

I can see the profile section is rendered in header.jsx and various environment variables like @account are available.

I see the render gets called from web_app where I dial into the API data and I used javascript to send that data to the interface level. I can set the data in web_app:

```
    - @data = generate_data
    - Rails.logger.debug("generate_data generated: #{@data.inspect}")
```

I'm certain I should not use javascript to push that data, but web_app does not always seem to be the only thing firing the interface component header.jsx 

The Join Date updates with the @account at times when my web_app code is not seemingly fired.

Has anyone suggestions or are there resources for getting started extending your mastodon instance I'd appreciate!

thank you!

**Comment by ThisIsMissEm** [comment]
Do you actually need to extend Mastodon here, or could you have the external service update data on the account via a Mastodon API integration? Or via providing verified URLs?

That is to say, the modifications you make to your mastodon server will only be usable on your mastodon server, and not via others, if done in a non-standard way. Perhaps you can achieve what you want through webhooks and automations?

**Reply by jtayler** [reply]
I'm not sure but I do not think I can do it with web hooks as I need to update the interface somewhat. 
![verifiedcloseup](https://github.com/mastodon/mastodon/assets/1004984/8c3749bb-3f34-4ba1-9a0d-f9e09332c5ee)
This rank and score there tabulates confidence of say, GitHub+LinkedIn+Medium —but without always having to display these accounts on our profile and in a way that easy to confirm and portable across whatever instances you choose to participate.
Does that help you see what needs being done here? Thank you!

**Reply by ThisIsMissEm** [reply]
That's not really how Mastodon's verification system is designed to work, and it also won't federate as you expect.

**Reply by jtayler** [reply]
you're welcome to try it along with us -- it doesn't interfere but extends the same verify, it's actually been around...

**Reply by thejennglass** [reply]
It really is a great way to help and it’s much easier to integrate than many other systems!

**Reply by fleckc3** [reply]
Same, I've used this before as well. You basically boost your own rank/score with links and profiles which is literally the same “rel=me” scheme Mastodon uses already, just easier and more portable.

**Reply by ThisIsMissEm** [reply]
Okay, but this isn't something Mastodon supports, so it won't necessarily work across all clients — you could certainly build this as a "link in bio" page that requires verified links and links back to the mastodon account, acting as a bridge + information for how many verified links are present.

But I'm not sure how you intend this to work in a federated setting; it's totally fine if you choose not to make it work in a federated way, but would mean you give different experiences to local vs remote users, so that's worth thinking about.

mastodon did once have cryptographic based verification, backed by keybase or the like, but this was removed because for most users it was just too complicated (citation needed, that's my memory of it)

**Reply by jtayler** [reply]
It's been around since Mastodon started to support the rel=me tag - it just makes verify easier, more private and portable across instances. This person is using it for Mastodon: https://mastodon.social/@mikehill@infosec.exchange

**Reply by jtayler** [reply]
<img width="588" alt="Screenshot 2023-12-14 at 10 20 49 AM" src="https://github.com/mastodon/mastodon/assets/1004984/9de5bf5b-b003-402c-8078-3bf9b4499777">

This supports the rel=me tag so it automatically lights up on Mastodon without you having to edit HTML

**Reply by thejennglass** [reply]
![image](https://github.com/mastodon/mastodon/assets/153785552/26dd6c82-2b49-420d-bce5-e06bea291974)
I'm using it too to support my Mastodon account

**Reply by ThisIsMissEm** [reply]
Okay, so, I'm really confused why exactly you're needing to edit the code of Mastodon, when the link already verifies and you could just click the link to see the third party data about the "verification" of the account. Truanon just seems to be another service like KeyOxide, Keybase, etc.

In theory any object with the "verified_at" timestamp would potentially show as verified, though it's unclear exactly if this would work 100% as expected across different clients.

**Reply by jtayler** [reply]
basically, we're implementing API to make things easier and to implement the privacy controls.

the link list gets called from _web_app haml and rendered in the jsx header component but I think React means the controller is not always fired while the @account object is changed?

the result is the new profile can be updated in some circumstances while displaying stale data. A simple refresh works to clear that.

how should server data be pushed to the client side so that the data remains fresh? (now it's using javascript which cannot be right)

I suspect this is just a lack of understanding how react works with the data model or how mastodon architecture handles data.

**Reply by ThisIsMissEm** [reply]
Data is fetched by redux, rendered by react & served by the API routes. But once again, any changes you make here risk breaking your compatibility with mobile clients & federating servers.

**Reply by jtayler** [reply]
thanks, that's super helpful in that I can research or perhaps simply sharing my screen would reveal what I've got wrong here —I'm certainly not passing the data around correctly.

in consideration of mobile clients or federated servers; I can't see how it could interfere? it doesn't change the service, there's no effect in its absence. is there something here that we should watch out for?

**Reply by ThisIsMissEm** [reply]
If you want me to pair with you, you're gonna have to hire me (sorry, just how it is in this capitalist world of ours)

Well, they might expect only links to contain the verified_at timestamp? Maybe they'll do weird things if that's not the case? Maybe it'll confuse users across the fediverse if they can't see the link easily that's been verified? Point is, I'd suggest caution here, and ask if the value vs potential incompatibility trade off is worth it.

**Reply by jtayler** [reply]
Ah yes I see what you mean to watch out for -- think of this case as a weather API: we display the temp. and if that doesn't happen it just doesn't show up

Of course, you are a valued expert! 

Mastodon, and the local artist group integrating this are non-commercial and so this API service is free but software engineering is not -- in this case, I offered to implement the API or at least get a healthy head-start because I think it's interesting. I have it running here at my desk and it's fun and although I use Mastodon I never tried to program it!

If we know what's involved maybe we can figure funding support --

I wrote a ruby class as a service I call from the _web_app haml file

```
TruAnonService.new(@account).get_profile

```

Now, I realize this was not the way I see anyone else move data, but I was able to pass data to the interface via javascript 

```
:javascript
  var verified_render_text = '#{@verified_render}';

```
there are times the header.jsx interface is updated with a new @account object but the _web_app haml is not fired again, thus leaving stale data.

I did not figure how model objects can be bound to interface context and this is likely the redux design pattern you describe.

The APIs have been implemented in a lot of languages and places and never seems to be a complicated affair so I'm hoping this is just a matter of proper implementation.

what do you think is involved?

**Reply by jtayler** [reply]
I created a route to bridge the jsx using an application controller that simply fetches and returns processed data. The page iterates and does final rendering and this seems to be the proper implementation design and handles the data as expected.

I notice the rendering is called three times, that's likely expected -- and only the last time does this data come in and display and I'm not sure when that is expected.

This is "ok" but makes the rendering decidedly after the rest of the region and I wonder why?

**Comment by kevinh21** [comment]
I’ve used this before and for me it’s easier than editing HTML.  Also, it would be portable across instances and I find that useful.

**Reply by ThisIsMissEm** [reply]
Used what? @jtayler didn't mention what he was trying to do in the OP, and only later mentioned something about Truanon. I'm not sure what you're replying to here @kevinh21 ?

**Reply by AngelM777** [reply]
Hey ThisIsMissEm, just looked at your profile and love your tech startup. I support it 💓
